core/terminal.go

```diff
@@ -52,41 +29,73 @@ type TerminalArgs struct {
 	InsecureRootCapabilities *bool `default:"false"`
 }
 
-func (container *Container) Terminal(svcID *call.ID, args *TerminalArgs) (*Terminal, http.Handler, error) {
-	termID := svcID.Digest()
-	endpoint := "terminals/" + termID.Encoded()
-	term := &Terminal{Endpoint: endpoint}
-	return term, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		var upgrader = websocket.Upgrader{}
-		ws, err := upgrader.Upgrade(w, r, nil)
+func (container *Container) Terminal(
+	ctx context.Context,
+	svcID *call.ID,
+	args *TerminalArgs,
+) error {
+	container = container.Clone()
+
+	container, err := container.WithExec(ctx, ContainerExecOpts{
+		Args:                          args.Cmd,
+		SkipEntrypoint:                true,
+		ExperimentalPrivilegedNesting: *args.ExperimentalPrivilegedNesting,
+		InsecureRootCapabilities:      *args.InsecureRootCapabilities,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create container for interactive terminal: %w", err)
+	}
+
+	return container.terminal(ctx, svcID, args)
+}
+
+func (dir *Directory) Terminal(
+	ctx context.Context,
+	svcID *call.ID,
+	ctr *Container,
+	args *TerminalArgs,
+) error {
+	var err error
+
+	if ctr == nil {
+		ctr, err = NewContainer(dir.Query, dir.Platform)
 		if err != nil {
-			bklog.G(r.Context()).WithError(err).Error("terminal handler failed to upgrade")
-			w.WriteHeader(http.StatusInternalServerError)
-			return
+			return err
```

Comment:
nit: here and elsewhere, I've been trying to be more consistent about always wrapping/annotating error messages, just because it helps trace down exactly where they are coming from when they get hit rather than having to guess